File: javascript/tests/before-refactor/PromotionTextDecorator.test.js
const { PromotionTextDecorator } = require('../../before-refactor/PromotionTextDecorator');

describe('#decorate before refactoring', () => {
  describe('Given a product that is not promoted', () => {
    describe('When the product has no free shipping nor discount', () => {
      it('Should generate a promotion description identical to its description', () => {
        const description = 'Pre-owned vase';
        const product = {
          promoted: false,
          oldCentPrice: 10000,
          centPrice: 10000,
          freeShipping: false,
          description,
          promotionText: 'Some old text',
        };

        const promotionText = PromotionTextDecorator.decorate(product);

        expect(promotionText).toEqual(description);
        expect(product.promotionText).toEqual(promotionText);
      });
    });
    describe('When the product has free shipping and discount', () => {
      it('Should still generate a promotion description identical to its description', () => {
        const description = 'Pre-owned vase';
        const product = {
          promoted: false,
          oldCentPrice: 10000,
          centPrice: 9000,
          freeShipping: true,
          description,
          promotionText: 'Some old text',
        };

        const promotionText = PromotionTextDecorator.decorate(product);

        expect(promotionText).toEqual(description);
        expect(product.promotionText).toEqual(promotionText);
      });
    });
  });
  describe('Given a product that is promoted', () => {
    describe('When the product has free shipping but no discount', () => {
      it('Should generate a promotion description with free shipping text', () => {
        const description = 'Pre-owned vase';
        const product = {
          promoted: true,
          hideDiscount: false,
          oldCentPrice: 10000,
          centPrice: 10000,
          freeShipping: true,
          description,
          promotionText: 'Some old text',
        };

        const promotionText = PromotionTextDecorator.decorate(product);

        const expected = `Free shipping!\n${description}\nFree shipping!`;
        expect(promotionText).toEqual(expected);
        expect(product.promotionText).toEqual(promotionText);
      });
    });
    describe('When the product is discounted, does not hide discount and has no free shipping', () => {
      it('Should generate a promotion description with discount text', () => {
        const description = 'Pre-owned vase';
        const product = {
          promoted: true,
          hideDiscount: false,
          oldCentPrice: 10000,
          centPrice: 9000,
          freeShipping: false,
          description,
          promotionText: 'Some old text',
        };

        const promotionText = PromotionTextDecorator.decorate(product);

        const offInDollar = (product.oldCentPrice - product.centPrice) / 100;
        const expected = `Don't miss the deal, $${offInDollar} off!\n${description}\nDon't miss the deal, $${offInDollar} off!`;
        expect(promotionText).toEqual(expected);
        expect(product.promotionText).toEqual(promotionText);
      });
    });
    describe('When the product is discounted and has no free shipping but hides discount', () => {
      it('Should generate a promotion description with discount text', () => {
        const description = 'Pre-owned vase';
        const product = {
          promoted: true,
          hideDiscount: true,
          oldCentPrice: 10000,
          centPrice: 9000,
          freeShipping: false,
          description,
          promotionText: 'Some old text',
        };

        const promotionText = PromotionTextDecorator.decorate(product);

        expect(promotionText).toEqual(description);
        expect(product.promotionText).toEqual(promotionText);
      });
    });
    describe('When the product has free shipping and discount', () => {
      it('Should generate a promotion description with both free shipping and discount texts', () => {
        const description = 'Pre-owned vase';
        const product = {
          promoted: true,
          hideDiscount: false,
          oldCentPrice: 10000,
          centPrice: 9000,
          freeShipping: true,
          description,
          promotionText: 'Some old text',
        };

        const promotionText = PromotionTextDecorator.decorate(product);

        const offInDollar = (product.oldCentPrice - product.centPrice) / 100;
        const expected = `Don't miss the deal, $${offInDollar} off!\nFree shipping!\n${description}\nFree shipping!\nDon't miss the deal, $${offInDollar} off!`;
        expect(promotionText).toEqual(expected);
        expect(product.promotionText).toEqual(promotionText);
      });
    });
    describe('When the product has no free shipping and no discount', () => {
      it('Should generate a promotion description identical to its description', () => {
        const description = 'Pre-owned vase';
        const product = {
          promoted: true,
          hideDiscount: false,
          oldCentPrice: 10000,
          centPrice: 10000,
          freeShipping: false,
          description,
          promotionText: 'Some old text',
        };

        const promotionText = PromotionTextDecorator.decorate(product);

        expect(promotionText).toEqual(description);
        expect(product.promotionText).toEqual(promotionText);
      });
    });
  });
});

File: java/src/test/java/promotiontext/before/PromotionTextDecoratorTest.java
package promotiontext.before;

import static org.testng.Assert.assertEquals;

import org.testng.annotations.Test;
import promotiontext.model.Product;


/**
 * Tests for PromotionTextDecorator before refactoring. You should not modify this class when
 * completing the kata
 */
public class PromotionTextDecoratorTest {
  @Test
  public void constructor() {
    // for 100% coverage
    new PromotionTextDecorator();
  }

  @Test
  public void decorate_notPromotedNoFreeShippingNoDiscount() {
    String description = "Pre-owned vase";
    Product product = new Product();
    product.setPromoted(false);
    product.setOldCentPrice(10000);
    product.setCentPrice(10000);
    product.setFreeShipping(false);
    product.setDescription(description);
    product.setPromotionText("Some old text");

    String promotionText = PromotionTextDecorator.decorate(product);

    assertEquals(promotionText, description,
        "Returned promotion text should be identical to its description");
    assertEquals(product.getPromotionText(), description,
        "product's promotion text should be identical to its description");
  }

  @Test
  public void decorate_notPromotedFreeShippingAndNoHideDiscount() {
    String description = "Pre-owned vase";
    Product product = new Product();
    product.setPromoted(false);
    product.setOldCentPrice(10000);
    product.setCentPrice(9000);
    product.setFreeShipping(true);
    product.setHideDiscount(false);
    product.setDescription(description);
    product.setPromotionText("Some old text");

    String promotionText = PromotionTextDecorator.decorate(product);

    assertEquals(promotionText, description,
        "Returned promotion text should be identical to its description");
    assertEquals(product.getPromotionText(), description,
        "product's promotion text should be identical to its description");
  }

  @Test
  public void decorate_promotedFreeShippingNoDiscount() {
    String description = "Pre-owned vase";
    Product product = new Product();
    product.setPromoted(true);
    product.setOldCentPrice(10000);
    product.setCentPrice(10000);
    product.setFreeShipping(true);
    product.setDescription(description);
    product.setPromotionText("Some old text");

    String promotionText = PromotionTextDecorator.decorate(product);

    String expected = String.format("Free shipping!\n%s\nFree shipping!", description);
    assertEquals(promotionText, expected,
        "Returned promotion text should have free shipping header and footer");
    assertEquals(product.getPromotionText(), expected,
        "product's promotion text should have free shipping header and footer");
  }

  @Test
  public void decorate_promotedNoFreeShippingNoHideDiscount() {
    String description = "Pre-owned vase";
    Product product = new Product();
    product.setPromoted(true);
    product.setOldCentPrice(10000);
    product.setCentPrice(9000);
    product.setFreeShipping(false);
    product.setDescription(description);
    product.setHideDiscount(false);
    product.setPromotionText("Some old text");

    String promotionText = PromotionTextDecorator.decorate(product);

    int offInDollar = (product.getOldCentPrice() - product.getCentPrice()) / 100;
    String expected =
        String.format("Don't miss the deal, $%d off!\n%s\nDon't miss the deal, $%d off!",
            offInDollar, description, offInDollar);
    assertEquals(promotionText, expected,
        "Returned promotion text should have discount header and footer");
    assertEquals(product.getPromotionText(), expected,
        "product's promotion text should have discount header and footer");
  }

  @Test
  public void decorate_promotedFreeShippingNoHideDiscount() {
    String description = "Pre-owned vase";
    Product product = new Product();
    product.setPromoted(true);
    product.setOldCentPrice(10000);
    product.setCentPrice(9000);
    product.setFreeShipping(true);
    product.setDescription(description);
    product.setHideDiscount(false);
    product.setPromotionText("Some old text");

    String promotionText = PromotionTextDecorator.decorate(product);

    int offInDollar = (product.getOldCentPrice() - product.getCentPrice()) / 100;
    String offText = String.format("Don't miss the deal, $%d off!", offInDollar);
    String freeShippingText = "Free shipping!";
    String expected =
        String.join("\n", offText, freeShippingText, description, freeShippingText, offText);
    assertEquals(promotionText, expected,
        "Returned promotion text should have both discount and free shipping headers and footers");
    assertEquals(product.getPromotionText(), expected,
        "product's promotion text should have both discount and free shipping headers and footers");
  }

  @Test
  public void decorate_promotedNoFreeShippingHideDiscount() {
    String description = "Pre-owned vase";
    Product product = new Product();
    product.setPromoted(true);
    product.setOldCentPrice(10000);
    product.setCentPrice(9000);
    product.setFreeShipping(false);
    product.setDescription(description);
    product.setHideDiscount(true);
    product.setPromotionText("Some old text");

    String promotionText = PromotionTextDecorator.decorate(product);

    assertEquals(promotionText, description,
        "Returned promotion text should be identical to its description");
    assertEquals(product.getPromotionText(), description,
        "product's promotion text should be identical to its description");
  }

  @Test
  public void decorate_promotedNoFreeShippingNoDiscount() {
    String description = "Pre-owned vase";
    Product product = new Product();
    product.setPromoted(true);
    product.setOldCentPrice(10000);
    product.setCentPrice(10000);
    product.setFreeShipping(false);
    product.setDescription(description);
    product.setPromotionText("Some old text");

    String promotionText = PromotionTextDecorator.decorate(product);

    assertEquals(promotionText, description,
        "Returned promotion text should be identical to its description");
    assertEquals(product.getPromotionText(), description,
        "product's promotion text should be identical to its description");
  }
}
